pkg/controller/volume/persistentvolume/scheduler_binder_test.go

```diff
@@ -1302,63 +1431,110 @@ func TestCheckBindings(t *testing.T) {
+			delayFunc: func(t *testing.T, testEnv *testEnv, pod *v1.Pod, pvs []*v1.PersistentVolume, pvcs []*v1.PersistentVolumeClaim) {
+				pvc := pvcs[0]
+				pv := pvs[0]
 				// Update PVC to be fully bound to PV
-				newPVC := binding.pvc.DeepCopy()
-				newPVC.ResourceVersion = "100"
-				newPVC.Spec.VolumeName = binding.pv.Name
+				newPVC := pvc.DeepCopy()
+				newPVC.Spec.VolumeName = pv.Name
 				metav1.SetMetaDataAnnotation(&newPVC.ObjectMeta, annBindCompleted, "yes")
-
-				// Update pvc cache, fake client doesn't invoke informers
-				internalBinder, ok := testEnv.binder.(*volumeBinder)
-				if !ok {
-					t.Fatalf("Failed to convert to internal binder")
+				testEnv.client.CoreV1().PersistentVolumeClaims(newPVC.Namespace).Update(newPVC)
+			},
+		},
+		"bound-by-pv-controller-before-bind": {
+			initPVs:    []*v1.PersistentVolume{pvNode1a},
+			initPVCs:   []*v1.PersistentVolumeClaim{unboundPVC},
+			binding:    makeBinding(unboundPVC, pvNode1aBound),
+			apiPV:      pvNode1aBound,
+			apiPVC:     boundPVCNode1a,
+			shouldFail: true, // bindAPIUpdate will fail because API conflict
+		},
+		"binding-static-pv-succeeds-after-time": {
+			binding:  makeBinding(unboundPVC, pvNode1aBound),
+			initPVs:  []*v1.PersistentVolume{pvNode1a},
+			initPVCs: []*v1.PersistentVolumeClaim{unboundPVC},
+			delayFunc: func(t *testing.T, testEnv *testEnv, pod *v1.Pod, pvs []*v1.PersistentVolume, pvcs []*v1.PersistentVolumeClaim) {
+				pvc := pvcs[0]
+				pv := pvs[0]
+				// Update PVC to be fully bound to PV
+				newPVC := pvc.DeepCopy()
+				newPVC.Spec.VolumeName = pv.Name
+				metav1.SetMetaDataAnnotation(&newPVC.ObjectMeta, annBindCompleted, "yes")
+				if _, err := testEnv.client.CoreV1().PersistentVolumeClaims(newPVC.Namespace).Update(newPVC); err != nil {
+					t.Errorf("failed to update PVC %q: %v", newPVC.Name, err)
 				}
-
-				pvcCache := internalBinder.pvcCache
-				internalPVCCache, ok := pvcCache.(*pvcAssumeCache)
-				if !ok {
-					t.Fatalf("Failed to convert to internal PVC cache")
+			},
+		},
+		"binding-dynamic-pv-succeeds-after-time": {
+			claimToProvision: pvcSetSelectedNode(provisionedPVC, "node1"),
+			initPVCs:         []*v1.PersistentVolumeClaim{provisionedPVC},
+			delayFunc: func(t *testing.T, testEnv *testEnv, pod *v1.Pod, pvs []*v1.PersistentVolume, pvcs []*v1.PersistentVolumeClaim) {
+				pvc := pvcs[0]
+				// Update PVC to be fully bound to PV
+				newPVC, err := testEnv.client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Get(pvc.Name, metav1.GetOptions{})
+				if err != nil {
+					t.Errorf("failed to get PVC %q: %v", pvc.Name, err)
+					return
+				}
+				dynamicPV := makeTestPV("dynamic-pv", "node1", "1G", "1", newPVC, waitClass)
+				dynamicPV, err = testEnv.client.CoreV1().PersistentVolumes().Create(dynamicPV)
+				if err != nil {
+					t.Errorf("failed to create PV %q: %v", dynamicPV.Name, err)
+					return
+				}
+				newPVC.Spec.VolumeName = dynamicPV.Name
```

Comment:
does this need to be a copy of the pvc object?

Reply:
No, `newPVC` is returned by `client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Get` which retrieves objects from apiserver. I've updated volumeReactor to return a copy of stored object.

Reply:
For dynamic provisioning, PVC was updated in bindAPIUpdate, this is why I need to fetch PVC from apiserver again.